File: src/store/deltaLog.ts
// Log-structured delta persistence (TRDD-K3WDPR7M Phase 4) — stop rewriting whole files to record a
// few KB of change.
//
// THE PROBLEM, VERIFIED FROM THE CODE AND THE DISK (not from memory):
//   log-sessions.json  31.6 MB rewritten every 300 s  =>  6.3 MB/min
//   log-offsets.json    3.1 MB rewritten every  60 s  =>  3.1 MB/min
// ~9.4 MB/min of device writes, forever, to persist a handful of changed records. On an SSD that is
// pure wear: the same blocks, rewritten thousands of times a day.
//
// THE FIX: append only what CHANGED, and rebuild state as snapshot + replayed deltas. A turn that
// touches one session appends a few hundred bytes instead of 31.6 MB.
//
// WHY NOT JUST TRACK DIRTY FLAGS UPSTREAM: the server only knows "something changed" (one boolean for
// the entire card collection), and threading per-record dirty tracking through every mutation site is
// exactly the kind of invariant that silently rots — one missed mutation and a card is never persisted
// again, which is a data-loss bug you would not notice for weeks. Hashing each record's serialized
// form is O(size) CPU but ZERO extra device writes, and it CANNOT miss a change: what is written is
// derived from what is actually there, not from a flag someone remembered to set.
import * as fs from 'fs'
import * as path from 'path'
import { createHash } from 'crypto'

/** Compact when the delta log grows past this multiple of the snapshot — bounded replay cost on load,
 *  while still amortizing the (expensive) full rewrite over many cheap appends. */
export const COMPACT_RATIO = 1.0
/** ...but always allow a small delta before compacting, so a tiny snapshot doesn't force a rewrite on
 *  every single append (a 2 KB snapshot with a 2 KB delta would otherwise compact constantly). */
export const COMPACT_MIN_BYTES = 256 * 1024

export interface SaveResult {
  /** Records appended (changed or new) this save. */
  appended: number
  /** Tombstones appended for records that disappeared. */
  deleted: number
  /** Bytes actually written to the device — the number that matters. */
  bytes: number
  compacted: boolean
}

interface Line<T> { k: string; v?: T; d?: 1 }

function sha(s: string): string {
  return createHash('sha256').update(s).digest('hex')
}

/**
 * A keyed record collection persisted as `<name>.snapshot.ndjson` + `<name>.delta.ndjson`.
 *
 * Durability: an append is a single `appendFileSync` of complete lines. A crash mid-append can leave a
 * TRAILING PARTIAL LINE, so load() drops a final unparseable line rather than throwing — losing the
 * last few hundred bytes of a delta is recoverable (the record is re-derived and re-appended on the
 * next save), whereas refusing to start because of it would strand the whole corpus.
 */
export class DeltaLog<T> {
  private readonly snapshotPath: string
  private readonly deltaPath: string
  /** key -> sha of the last SERIALIZED form we persisted. This is what makes a save a no-op when
   *  nothing changed — without it we would re-append every record every interval and gain nothing. */
  private written = new Map<string, string>()

  constructor(dir: string, name: string) {
    this.snapshotPath = path.join(dir, `${name}.snapshot.ndjson`)
    this.deltaPath = path.join(dir, `${name}.delta.ndjson`)
  }

  /** Read NDJSON, skipping blank lines and tolerating ONE trailing partial line (a torn append). */
  private readLines(file: string): Array<Line<T>> {
    let raw: string
    try { raw = fs.readFileSync(file, 'utf8') } catch { return [] }
    const lines = raw.split('\n')
    const out: Array<Line<T>> = []
    for (const [i, ln] of lines.entries()) {
      if (ln.trim() === '') continue
      try {
        out.push(JSON.parse(ln) as Line<T>)
      } catch {
        // Only the LAST line may be torn (an append is atomic per write, and a crash can only cut the
        // final one). A corrupt line anywhere else means real damage — say so instead of silently
        // dropping records the caller believes are persisted.
        if (i === lines.length - 1 || lines.slice(i + 1).every((l) => l.trim() === '')) break
        throw new Error(`${file}: corrupt line ${i + 1} (not a torn tail — the file is damaged)`)
      }
    }
    return out
  }

  /** Replay NDJSON files in order into a fresh map — the ONE definition of "how lines become state",
   *  shared by load() and compaction's verify so there is never a second, drifting parser. Later lines
   *  win; a tombstone (d:1) deletes. */
  private replay(files: string[]): Map<string, T> {
    const map = new Map<string, T>()
    for (const file of files) {
      for (const l of this.readLines(file)) {
        if (l.d === 1) map.delete(l.k)
        else if (l.v !== undefined) map.set(l.k, l.v)
      }
    }
    return map
  }

  /** Snapshot, then deltas replayed in order. Later lines win; tombstones delete. */
  load(): Map<string, T> {
    const map = this.replay([this.snapshotPath, this.deltaPath])

    // Seed the written-hashes from what we just loaded, or the first save after a restart would
    // re-append every record (turning a restart into a full 31.6 MB rewrite — the very thing we are
    // removing).
    this.written = new Map([...map].map(([k, v]) => [k, sha(JSON.stringify(v))]))
    return map
  }

  private sizeOf(file: string): number {
    try { return fs.statSync(file).size } catch { return 0 }
  }

  /** Test seam: called after the candidate snapshot bytes are written to `path` and BEFORE they are
   *  verified. Production no-op. A test overrides it to simulate a silently corrupted/truncated
   *  snapshot write and prove compaction's verify catches it and KEEPS the delta (TRDD-K3WDPR7M). */
  protected onSnapshotWritten(_path: string): void { /* no-op in production */ }

  /** Rewrite the snapshot from `records`, VERIFY it from disk, then drop the delta. The ONE expensive
   *  write — amortized. */
  private compact(records: Map<string, T>): number {
    let body = ''
    for (const [k, v] of records) body += JSON.stringify({ k, v }) + '\n'
    // temp + rename: a crash mid-compaction must never leave a half-written snapshot, which would
    // silently truncate the corpus on the next load.
    const tmp = `${this.snapshotPath}.tmp`
    fs.writeFileSync(tmp, body)
    this.onSnapshotWritten(tmp)

    // TRDD-K3WDPR7M (2026-07-15 USER directive: a source file may be deleted ONLY after the durable
    // destination is confirmed to hold ALL its data). The delta is the ONLY other copy of every change
    // it holds; a silently truncated/short snapshot write (disk full, fs bug) followed by rmSync would
    // destroy those records forever. So PROVE the snapshot before committing to it: read the candidate
    // BACK FROM DISK (not in-memory state — the whole point is to catch a bad write) and confirm it
    // reproduces EXACTLY the saved record set (count + every per-record hash). Verify the temp file
    // BEFORE the rename, so a failure leaves the OLD snapshot AND the delta fully intact — nothing is
    // committed, nothing is lost — and surface the fault per this file's fail-fast convention (throw,
    // exactly as readLines() throws on a corrupt mid-file line).
    const reloaded = this.replay([tmp])
    const expected = new Map<string, string>()
    for (const [k, v] of records) expected.set(k, sha(JSON.stringify(v)))
    let mismatch: string | null = null
    if (reloaded.size !== expected.size) {
      mismatch = `holds ${reloaded.size} records, expected ${expected.size}`
    } else {
      for (const [k, want] of expected) {
        const got = reloaded.get(k)
        if (got === undefined || sha(JSON.stringify(got)) !== want) {
          mismatch = `record ${JSON.stringify(k)} is not durable in the snapshot`
          break
        }
      }
    }
    if (mismatch !== null) {
      try { fs.rmSync(tmp, { force: true }) } catch { /* discard the bad candidate */ }
      throw new Error(`${this.snapshotPath}: compaction verify failed (${mismatch}); delta KEPT, snapshot NOT replaced — no records lost`)
    }

    fs.renameSync(tmp, this.snapshotPath)
    // Only AFTER the snapshot is durable AND verified — deleting the delta any earlier would lose every
    // change it holds if the rename never happened or the write was short.
    try { fs.rmSync(this.deltaPath, { force: true }) } catch { /* nothing to drop */ }
    return Buffer.byteLength(body)
  }

  /**
   * Persist `records`, writing ONLY what changed. Returns what it actually cost the device.
   *
   * Nothing changed ⇒ ZERO bytes written. That is the point: the current code writes 31.6 MB every
   * 5 minutes whether or not anything happened.
   */
  save(records: Map<string, T>): SaveResult {
    fs.mkdirSync(path.dirname(this.snapshotPath), { recursive: true })

    let body = ''
    let appended = 0
    let deleted = 0
    for (const [k, v] of records) {
      const json = JSON.stringify(v)
      if (this.written.get(k) === sha(json)) continue // unchanged — the whole saving
      body += JSON.stringify({ k, v }) + '\n'
      this.written.set(k, sha(json))
      appended++
    }
    for (const k of [...this.written.keys()]) {
      if (records.has(k)) continue
      body += JSON.stringify({ k, d: 1 }) + '\n' // tombstone: the record is gone
      this.written.delete(k)
      deleted++
    }
    if (body === '') return { appended: 0, deleted: 0, bytes: 0, compacted: false }

    const snapBytes = this.sizeOf(this.snapshotPath)
    const deltaBytes = this.sizeOf(this.deltaPath) + Buffer.byteLength(body)
    if (deltaBytes > COMPACT_MIN_BYTES && deltaBytes > snapBytes * COMPACT_RATIO) {
      return { appended, deleted, bytes: this.compact(records), compacted: true }
    }
    fs.appendFileSync(this.deltaPath, body)
    return { appended, deleted, bytes: Buffer.byteLength(body), compacted: false }
  }

  /** Total bytes on disk (both files) — for the persist stats the dashboard shows. */
  diskBytes(): number {
    return this.sizeOf(this.snapshotPath) + this.sizeOf(this.deltaPath)
  }
}

File: rust-core/crates/agentlens-core/src/delta_log.rs
//! Port of src/store/deltaLog.ts (TRDD-K3WDPR7M) for the P5e durable state (TRDD-DMWOBWFH):
//! a keyed record collection persisted as `<name>.snapshot.ndjson` + `<name>.delta.ndjson`,
//! appending ONLY what changed. The on-disk format is BYTE-COMPATIBLE with the TS engine —
//! `{"k":key,"v":record}` / `{"k":key,"d":1}` lines, later lines win, a tombstone deletes — so a
//! cutover restart reads either engine's files. (The per-key "last written" hash is in-memory
//! only, so it need not match TS's sha256; a 64-bit std hash of the serialized record is enough.)
//!
//! Durability rules carried over verbatim: an append is one write of complete lines, so load()
//! tolerates ONE torn TRAILING line and REFUSES a corrupt line anywhere else (that is real damage,
//! not a torn tail); compaction writes `<snapshot>.tmp`, re-reads it FROM DISK and verifies count +
//! every record before the rename, and deletes the delta only after the rename — a short or
//! corrupt snapshot write keeps the old snapshot AND the delta intact.

use std::collections::hash_map::DefaultHasher;
use std::fs;
use std::hash::{Hash, Hasher};
use std::io::Write;
use std::path::{Path, PathBuf};

use indexmap::IndexMap;
use serde_json::Value;

/// Compact when the delta grows past this multiple of the snapshot …
pub const COMPACT_RATIO: f64 = 1.0;
/// … but always allow a small delta first, so a tiny snapshot does not compact on every append.
pub const COMPACT_MIN_BYTES: u64 = 256 * 1024;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct SaveResult {
    pub appended: usize,
    pub deleted: usize,
    /// Bytes actually written to the device — the number that matters.
    pub bytes: u64,
    pub compacted: bool,
}

pub struct DeltaLog {
    snapshot_path: PathBuf,
    delta_path: PathBuf,
    /// key → hash of the last SERIALIZED form persisted; what makes an unchanged save a no-op.
    written: IndexMap<String, u64>,
}

fn hash_json(s: &str) -> u64 {
    let mut h = DefaultHasher::new();
    s.hash(&mut h);
    h.finish()
}

fn size_of(p: &Path) -> u64 {
    fs::metadata(p).map(|m| m.len()).unwrap_or(0)
}

impl DeltaLog {
    pub fn new(dir: &Path, name: &str) -> DeltaLog {
        DeltaLog {
            snapshot_path: dir.join(format!("{name}.snapshot.ndjson")),
            delta_path: dir.join(format!("{name}.delta.ndjson")),
            written: IndexMap::new(),
        }
    }

    /// Read one NDJSON file: blank lines skipped, ONE torn trailing line tolerated, a corrupt line
    /// anywhere else is an error (the file is damaged — say so instead of silently dropping
    /// records the caller believes are persisted). A missing file is empty.
    fn read_lines(file: &Path) -> Result<Vec<Value>, String> {
        let Ok(raw) = fs::read_to_string(file) else { return Ok(Vec::new()) };
        let lines: Vec<&str> = raw.split('\n').collect();
        let mut out = Vec::new();
        for (i, ln) in lines.iter().enumerate() {
            if ln.trim().is_empty() {
                continue;
            }
            match serde_json::from_str::<Value>(ln) {
                Ok(v) => out.push(v),
                Err(_) => {
                    if lines[i + 1..].iter().all(|l| l.trim().is_empty()) {
                        break;
                    }
                    return Err(format!("{}: corrupt line {} (not a torn tail — the file is damaged)", file.display(), i + 1));
                }
            }
        }
        Ok(out)
    }

    /// Replay files in order into a fresh map — the ONE definition of "how lines become state",
    /// shared by load() and compaction's verify.
    fn replay(files: &[&Path]) -> Result<IndexMap<String, Value>, String> {
        let mut map = IndexMap::new();
        for file in files {
            for l in Self::read_lines(file)? {
                let Some(k) = l.get("k").and_then(Value::as_str) else { continue };
                if l.get("d").and_then(Value::as_u64) == Some(1) {
                    map.shift_remove(k);
                } else if let Some(v) = l.get("v") {
                    map.insert(k.to_owned(), v.clone());
                }
            }
        }
        Ok(map)
    }

    /// Snapshot, then deltas, later lines win. Seeds the written-hashes so the first save after a
    /// restart does not re-append every record.
    pub fn load(&mut self) -> Result<IndexMap<String, Value>, String> {
        let map = Self::replay(&[&self.snapshot_path, &self.delta_path])?;
        self.written = map.iter().map(|(k, v)| (k.clone(), hash_json(&serde_json::to_string(v).unwrap_or_default()))).collect();
        Ok(map)
    }

    /// Rewrite the snapshot from `records`, VERIFY it from disk, then drop the delta.
    fn compact(&self, records: &IndexMap<String, Value>) -> Result<u64, String> {
        let mut body = String::new();
        for (k, v) in records {
            body.push_str(&serde_json::to_string(&serde_json::json!({ "k": k, "v": v })).unwrap_or_default());
            body.push('\n');
        }
        let tmp = PathBuf::from(format!("{}.tmp", self.snapshot_path.display()));
        fs::write(&tmp, &body).map_err(|e| format!("{}: {e}", tmp.display()))?;
        // PROVE the snapshot before committing to it: read the candidate BACK FROM DISK and confirm
        // it reproduces exactly the saved record set (count + every record).
        let reloaded = Self::replay(&[&tmp])?;
        let mismatch = if reloaded.len() != records.len() {
            Some(format!("holds {} records, expected {}", reloaded.len(), records.len()))
        } else {
            records.iter().find(|(k, v)| reloaded.get(*k) != Some(*v)).map(|(k, _)| format!("record {k:?} is not durable in the snapshot"))
        };
        if let Some(m) = mismatch {
            let _ = fs::remove_file(&tmp);
            return Err(format!("{}: compaction verify failed ({m}); delta KEPT, snapshot NOT replaced — no records lost", self.snapshot_path.display()));
        }
        fs::rename(&tmp, &self.snapshot_path).map_err(|e| format!("{}: rename: {e}", self.snapshot_path.display()))?;
        // Only AFTER the snapshot is durable AND verified.
        let _ = fs::remove_file(&self.delta_path);
        Ok(body.len() as u64)
    }

    /// Persist `records`, writing ONLY what changed. Nothing changed ⇒ ZERO bytes written.
    pub fn save(&mut self, records: &IndexMap<String, Value>) -> Result<SaveResult, String> {
        if let Some(dir) = self.snapshot_path.parent() {
            fs::create_dir_all(dir).map_err(|e| format!("{}: {e}", dir.display()))?;
        }
        let mut body = String::new();
        let mut appended = 0usize;
        let mut deleted = 0usize;
        for (k, v) in records {
            let json = serde_json::to_string(v).unwrap_or_default();
            let h = hash_json(&json);
            if self.written.get(k) == Some(&h) {
                continue; // unchanged — the whole saving
            }
            body.push_str(&format!("{{\"k\":{},\"v\":{json}}}\n", serde_json::to_string(k).unwrap_or_default()));
            self.written.insert(k.clone(), h);
            appended += 1;
        }
        let gone: Vec<String> = self.written.keys().filter(|k| !records.contains_key(*k)).cloned().collect();
        for k in gone {
            body.push_str(&format!("{{\"k\":{},\"d\":1}}\n", serde_json::to_string(&k).unwrap_or_default())); // tombstone
            self.written.shift_remove(&k);
            deleted += 1;
        }
        if body.is_empty() {
            return Ok(SaveResult::default());
        }
        let snap_bytes = size_of(&self.snapshot_path);
        let delta_bytes = size_of(&self.delta_path) + body.len() as u64;
        if delta_bytes > COMPACT_MIN_BYTES && delta_bytes as f64 > snap_bytes as f64 * COMPACT_RATIO {
            let bytes = self.compact(records)?;
            return Ok(SaveResult { appended, deleted, bytes, compacted: true });
        }
        let mut f = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.delta_path)
            .map_err(|e| format!("{}: {e}", self.delta_path.display()))?;
        f.write_all(body.as_bytes()).map_err(|e| format!("{}: {e}", self.delta_path.display()))?;
        Ok(SaveResult { appended, deleted, bytes: body.len() as u64, compacted: false })
    }

    /// Total bytes on disk (both files).
    pub fn disk_bytes(&self) -> u64 {
        size_of(&self.snapshot_path) + size_of(&self.delta_path)
    }
}
